### PR TITLE
armv7 images

### PR DESCRIPTION
**Pull request by Silex**
Hi!

Just wondering if you plan to add avmv7 alpine images.

### CODE REVIEW

**Comment by Silex** [comment]
Okay... I can build on arm64 using travis beta arm builds, but this makes the amd64 builds fail: https://travis-ci.org/Silex/docker-emacs/builds/633236712

This happens because to build Emacs < 27 you need to disable `randomize_va_space` (see https://github.com/moby/moby/issues/22801) and the image they use for multiarch is "too modern" :-)

Anyway, it looks like this would not support arm32 images and I'll need to create multi-arch manifests for each images anyway so I don't think building the images separatly makes sense.

I'll confirm and then try the qemu solution.

**Comment by NicolasPetton** [comment]
Philippe Vaucher <notifications@github.com> writes:

> Okay... I can build on arm64 using travis beta arm builds, but this makes the amd64 builds fail: https://travis-ci.org/Silex/docker-emacs/builds/633236712
>
> This happens because to build Emacs < 27 you need to disable `randomize_va_space` (see https://github.com/moby/moby/issues/22801).
>
> Anyway, it looks like this would not support arm32 images and I'll need to create multi-arch manifests for each images anyway so I don't think building the images separatly makes sense.
>
> I'll confirm and then try the qemu solution.

Thanks!

**Comment by Silex** [comment]
@NicolasPetton: can you test `docker run -it --rm silex/emacs:26.3-test` and `docker run -it --rm silex/emacs:26.3-dev-test` and report if they work correctly?

**Comment by Silex** [comment]
Travis has a limitation of 50 minutes per job. So that means one job cannot build the 4 kind of amd64/i386/armv7/arm64v8 because one of them takes ~15 minutes *natively*, but when crossbuilding through QEMU it takes ~45m for each arm images, resulting in ~2-3h per image.

So I'm now going for the "everything is split" route, see https://travis-ci.org/Silex/docker-emacs/builds/636042712. That's 168 job with a max of 200, so that will be a problem if when we add 27.1 27.2 27.3 28.1 28.2 etc, but then I can easily divide by two the number of jobs by building amd64/i386 together and armv7/arm64v8 together, which should reasonably take ~30-40 minutes and not exceed 50 minutes. I might even do that now for clarity reasons because looking at 168 jobs is a bit confusing when looking at the status.

**Comment by NicolasPetton** [comment]
@Silex Thanks for all effort you're putting in this!

Report an a RPI3 (armv7l kernel):

```
docker run -it --rm silex/emacs:26.3-test
Unable to find image 'silex/emacs:26.3-test' locally
26.3-test: Pulling from silex/emacs
b0b312c654a9: Already exists 
f7f8a2b7886d: Already exists 
2d07ab3b900d: Already exists 
fc1f49a58853: Already exists 
8a7e80ceb66e: Pull complete 
9aa5d2e3790a: Pull complete 
3681650494b3: Pull complete 
Digest: sha256:9b7a4ba2d29dbafe12bd98c2dbabaee4463f918804904e59b73fa50c0425f351
Status: Downloaded newer image for silex/emacs:26.3-test
```
I confirm that Emacs works just fine in the image!

**Comment by Silex** [comment]
Alright. I finally have all the informations necessary to finish this.

Basically there's 3 ways to do multiarch images:

1. Dockerfile contains `FROM $ARCH/ubuntu:18.04`, and you use `docker build --build-arg ARCH=arm32v7`.
1. Dockerfile contains `FROM ubuntu:18.04`, and you use `docker build --platform arm`.
1. Dockerfile contains `FROM ubuntu:18.04`, and you use `docker buildx --platforms linux/arm32v7`.

Method 1 is for old docker engines. Method 1 & 2 need manual manifest creation. Method 2 & 3 only work with new docker engines AND with experimental mode on. Thanks to https://github.com/moby/moby/issues/36552 I know now that method 2 only works if you `--pull` when you build. Method 3 would be the best choice if travis didn't timeout jobs after 50 minutes, as building 4 images in order to create one takes too long. This limitation could have been worked around with multiple jobs on different architectures, but I didn't find a way for a travis job to "connect" to another one and build a temporary building swarm.

So it looks like I have no choice but to embrace method 2 for now.

When docker has matured more maybe I'll switch to something else. I also accidently discovered I could possibly skip travis entirely and build on dockerhub, using the hooks like in https://github.com/rmoriz/multiarch-test.

Good post about difference bwteen 1 and 3: https://dev.to/jdrouet/multi-arch-images-with-docker-152f

**Comment by DamienCassou** [comment]
I think you once said you wanted to investigate gitlab and its docker support. I don't remember if you did though. Maybe it would be a good time?

**Comment by Silex** [comment]
@DamienCassou: I did, and yes it looks way nicer than travis, especially the local registry.

That said, gitlab free edition only offers 2000 CI minutes per month, which makes ~60 minute per day. It *might* just be enough but it's more likely that it is not, for example if a base image changes and especially now that we'll build multi arch images. When cross-building with qemu a single image can take 30-40 minutes to build.

The only CI platforms that seem "unlimited" for open source are travis and the docker hub.

**Comment by NicolasPetton** [comment]
Philippe Vaucher <notifications@github.com> writes:

> @DamienCassou: I did, and yes it looks way nicer than travis,
> especially the local registry.

Not sure if it can be useful or not, but I also have a DroneCI instance
running on a rpi3, that I could give you access to.

Nico

**Comment by Silex** [comment]
@NicolasPetton: ah, this is interesting!

I did a bit of more research:

| ci             | monthly (m) | daily (h) | job timeout | job max | arm | arm64 | amd64 |
|----------------|-------------|-----------|-------------|---------|-----|-------|-------|
| travis         | unlimited   | unlimited |          50 | 200     | no  | yes   | yes   |
| circleci       | 400000      | 215       |         300 | ?       | no? | no?   | yes   |
| cloud.drone.io | unlimited   | unlimited |        900? | ?       | yes | yes   | yes   |
| gitlab         | 50000       | 26        |         180 | ?       | no? | no?   | yes   |

Actually gitlab for opensource is like having the gold plan! So gitlab looks like a viable option!

@DamienCassou: I have trouble finding documentation about ARM support. I know I can crossbuild it all through QEMU and just use one runner with `docker builx`, so it looks awesome but still, building the arm images on an arm runner makes more sense.

Anyway, I'll play around with gitlab to get a better idea. Drone seems "simpler" but maybe a bit too simple, e.g you cannot register qemu so the buildx way is impossible? https://discourse.drone.io/t/run-docker-container-in-docker/5704

I'll try to get a basic gitlab buildx version going (like I originally had for travis).

---

**UPDATE** okay the gitlab shared runners actually have an unconfigurable 3h limit (https://gitlab.com/Silex777/docker-emacs/-/jobs/407495608 https://gitlab.com/gitlab-org/gitlab/commit/3f4a3e4f608?view=parallel). Which means I still cannot use method 3 :-/ And given that gitlab does not offer arm shared runners (correct me if I'm wrong) it looks like travis is still the best option out there. Cloud.drone.io would be nice *if* it was not so beta. I ran a "hello world" bash that just echoed strings and after 2h the job was still not started. I guess they lack a lot of runners. Currently testing gitlab building a single arm32v7 image to see if it takes more than 3h.

------

**UPDATE2** okay, gitlab crossbuilding to armv7 *almost* timeout, taking 179 minutes out of 180 to build the 26.3-dev and 26.3 image (https://gitlab.com/Silex777/docker-emacs/-/jobs/407668850). Image 26.3-dev alone takes 172 minutes (https://gitlab.com/Silex777/docker-emacs/-/jobs/407717304), so only 8 minutes spare. Maybe the solution will be to add my own runner, I have a box on digital ocean but it's really slow...

------

**UPDATE3** sigh... just found a new problem: I cannot find a way to disable ASLR on arm64. Tried everything, even started the docker daemon with `setarch -R` but no (https://travis-ci.org/Silex/docker-emacs/builds/638812299). That would mean I HAVE to crossbuild from amd64. If gitlab does not have jobs limitation I think I'll move there, in the meantime I asked there https://travis-ci.community/t/cannot-disable-aslr-in-arm64-jobs/6915.

**Comment by Silex** [comment]
FYI tests are continuing at https://gitlab.com/Silex777/docker-emacs

Basically what I found so far is that the public runners have a limit of 180 minutes and this often timeout when building armv7. Also the pipeline becomes incredibly complex :disappointed: 

So, I'm leveraging a box I have as a runner dedicated for this project, and that way I can use `buildx` and have a simple pipeline. Given the box is a 5$ digital ocean droplet it's taking ages but at least it works the way I want. Will keep you posted when I have something ready.

**Comment by DamienCassou** [comment]
thank you so much for all the energy you are putting into this.

-- 
Damien Cassou

"Success is the ability to go from one failure to another without
losing enthusiasm." --Winston Churchill

**Comment by Silex** [comment]
Just a quick update because it made me laugh: on this server of mine it took 531 minutes (~9h) (https://gitlab.com/Silex777/docker-emacs/-/jobs/414066069) to build **1** multi-platform image... given we have 42 images that cannot scale well.

I realised this box was really old and for the same cost I now have one that has 4 times the RAM and twice the processors, so I'm testing again to see if it's feasible, but I think it's likely that I'll have the amd/i386 variants built on the shared runners (by using tags), and the arm variants built on my runner.

It's quite some work but it's interesting, gitlab is much nicer than travis, and at the same time the fact that it lacks simple matrix-style declarations is a bit puzzling.

Also, if this becomes too much of a pain I'll just drop some architectures...

**Comment by Silex** [comment]
Fantastic help about buildx/qemu gotchas: https://nexus.eddiesinentropy.net/2020/01/12/Building-Multi-architecture-Docker-Images-With-Buildx/

Good gitlab example: https://github.com/hectorm/docker-supercronic/blob/master/.gitlab-ci.yml

**Comment by Silex** [comment]
QEMU bugs with alpine images: https://bugs.launchpad.net/qemu/+bug/1861161

**Comment by DamienCassou** [comment]
Philippe Vaucher <notifications@github.com> writes:
> QEMU bugs with alpine images: https://bugs.launchpad.net/qemu/+bug/1861161

:-(

-- 
Damien Cassou

"Success is the ability to go from one failure to another without
losing enthusiasm." --Winston Churchill

**Comment by Silex** [comment]
Just for information there are more informations at https://github.com/WorksOnArm/cluster/issues/192 (cloud.drone.io is a dead end because I cannot disable ASLR, still some hopes with the QEMU route but it's getting slim).

**Comment by Silex** [comment]
Sorry a bit swamped to make this move forward, but https://github.com/moby/qemu/issues/9 looks promising.

**Comment by Silex** [comment]
Good news! I finally was able to compile arm64 & arm32 images on a new server. Will keep you posted but theorically the hurdles are over.

**Comment by DamienCassou** [comment]
awesome news. Great job!

-- 
Damien Cassou

"Success is the ability to go from one failure to another without
losing enthusiasm." --Winston Churchill

**Comment by NicolasPetton** [comment]
Philippe Vaucher <notifications@github.com> writes:

> Good news! I finally was able to compile arm64 & arm32 images on a new
> server. Will keep you posted but theorically the hurdles are over.

Hi Philippe,

Does this mean that you do not need an arm32 host anymore?

Cheers,
Nico

**Comment by Silex** [comment]
> Does this mean that you do not need an arm32 host anymore?

Yes, thanks for the offer tho.

Shall I get out of sponsorship I'll think about you :wink:

**Comment by Silex** [comment]
@NicolasPetton, @DamienCassou: if you guys want to run some tests before I deploy like crazy, test one or more of these images: https://hub.docker.com/r/silex/emacs/tags?page=1&name=-test

Probably that `silex/emacs:26.3-dev-test` is the one that you are already using. As you can see it supports 4 architectures :tada: 

I'll still need a few more days to refactor all my tests into something production ready tho.

**Comment by Silex** [comment]
@NicolasPetton, @DamienCassou: very close to success, here's what I ran into:

- Ubuntu 12.04 is not multiarch, thus I had to upgrade a lot of images to Ubuntu 14.04, and thus I had to downgrade texinfo for some images because Emacs wouldn't build.
- Emacs 24.1 would not build on 14.04 (gcc version), I had to make a patch commenting a line in stdio.h

I have a few hurdles left (https://gitlab.com/Silex777/docker-emacs/pipelines/144236976):

- Emacs 23.4 simply refuses to build on aarch64:

```
checking build system type... build-aux/config.guess: unable to guess system type

This script, last modified 2006-06-06, has failed to recognize
the operating system you are using. It is advised that you
download the most up to date version of the config scripts from

  http://savannah.gnu.org/cgi-bin/viewcvs/*checkout*/config/config/config.guess
and
  http://savannah.gnu.org/cgi-bin/viewcvs/*checkout*/config/config/config.sub

If the version you run (build-aux/config.guess) is already up to date, please
send the following data and any information you think might be
pertinent to <config-patches@gnu.org> in order to provide the needed
information to handle your system.

config.guess timestamp = 2006-06-06

uname -m = aarch64
uname -r = 4.15.0-72-generic
uname -s = Linux
uname -v = #81-Ubuntu SMP Tue Nov 26 12:21:09 UTC 2019

/usr/bin/uname -p = 
/bin/uname -X     = 

hostinfo               = 
/bin/universe          = 
/usr/bin/arch -k       = 
/bin/arch              = 
/usr/bin/oslevel       = 
/usr/convex/getsysinfo = 

UNAME_MACHINE = aarch64
UNAME_RELEASE = 4.15.0-72-generic
UNAME_SYSTEM  = Linux
UNAME_VERSION = #81-Ubuntu SMP Tue Nov 26 12:21:09 UTC 2019
configure: error: cannot guess build type; you must specify one
```

If you have an idea of what I can try?

- The alpine version of 25.2 and 25.3 crash when building on aarch64:

```
 #16 498.1 make[2]: *** [Makefile:256: ../lisp/term/ns-win.elc] Killed
 #16 498.2 make[1]: *** [Makefile:725: ../lisp/term/ns-win.elc] Error 2
 #16 498.2 make[2]: Leaving directory '/opt/emacs/lisp'
 #16 498.2 make[1]: Leaving directory '/opt/emacs/src'
 #16 498.2 make: *** [Makefile:398: src] Error 2
```

Basically what happens is that one of the process (ones that has `(setq max-lisp-eval-depth 2200)` in it) is eats more and more memory until it gets killed. I think it's when byte compiling.

So, I'm not sure about what to do. I could simply not build all variants on for these images (e.g only amd64/i386), but then it's a bit unexpected when doing CI? Give me your thoughts.

**Comment by DamienCassou** [comment]
Philippe Vaucher <notifications@github.com> writes:
> So, I'm not sure about what to do. I could simply not build all
> variants on for these images (e.g only amd64/i386), but then it's a
> bit unexpected when doing CI? Give me your thoughts.

To me this is no surprise that some versions of Emacs can't be built on
some architecture or OS. I think this is the same for all software: as
new releases arrive, support for some architectures is added and support
for others is removed.

I would be perfectly happy with just a few Emacs versions if that
includes the latest.

-- 
Damien Cassou

"Success is the ability to go from one failure to another without
losing enthusiasm." --Winston Churchill

**Comment by Silex** [comment]
Ok I decided to build what builds for now, and I'll make issues about what doesn't build for later solving.

**Comment by Silex** [comment]
:tada:

More modifications will follow in #58 

Please test a bit and notice me if you find anything.

**Comment by DamienCassou** [comment]
Thank you so much for your work @Silex. For now, it doesn't seem to work but I'm not sure this has anything to do with your images:

https://drone.petton.fr/mpdel/libmpdel/73/1/2

Maybe related to https://github.com/drone/drone/issues/2955?

**Comment by Silex** [comment]
@DamienCassou: good catch, apparently something didn't build `latest` properly!

https://hub.docker.com/r/silex/emacs/tags?page=1&name=latest

As you see https://hub.docker.com/r/silex/emacs/tags?page=1&name=26.3 are multiarch.

I'll figure it out and report.

**Comment by Silex** [comment]
Ok it's because of https://github.com/docker/buildx/issues/177

I'll work around it in the meantime.

**Comment by Silex** [comment]
Hum in trying to fix it I also broke 26.3*.

The good news is that I think it'd be fixed soon and be multiarch for all tags :-)

**EDIT**: it's actually more complicated :-/

**Comment by Silex** [comment]
Okay, all tags are now multiarch but my workaround sucks (the tags don't point to exactly the same image digests yet the layers are identical). I'm trying to find something elegant.

**Comment by DamienCassou** [comment]
> Okay, all tags are now multiarch but my workaround sucks (the tags don't point to exactly the same image digests yet the layers are identical). I'm trying to find something elegant.

and it is working **great**!!!!

Before I had to wait at least 5 minutes for a build to finish. Now it's less than 1. Thank you so much.

**Comment by NicolasPetton** [comment]
Damien Cassou <notifications@github.com> writes:

> and it is working **great**!!!!

That's awesome! Thanks a lot!
